src/ddsrt/src/events/posix/events_posix.c

```diff
@@ -234,68 +223,99 @@ dds_return_t ddsrt_event_queue_wait(ddsrt_event_queue_t* queue, dds_duration_t r
     }
     ddsrt_mutex_unlock(&queue->lock);
   }
-  else {
+  else
+  {
     /*something else happened*/
     return DDS_RETCODE_ERROR;
   }
 
   return DDS_RETCODE_OK;
 }
 
-dds_return_t ddsrt_event_queue_signal(ddsrt_event_queue_t* queue) {
+dds_return_t ddsrt_event_queue_signal(ddsrt_event_queue_t* queue)
+{
 #if !defined(LWIP_SOCKET)
-  char buf = 0;
+  char buf = 0x0;
+  int n = 0;
 #if defined(_WIN32)
-  if (1 != send(queue->interrupt[1], &buf, 1, 0))
-    return DDS_RETCODE_ERROR;
+  n = send(queue->interrupt[1], &buf, 1, 0);
 #else
-  if (1 != write(queue->interrupt[1], &buf, 1))
+  n = (int)write(queue->interrupt[1], &buf, 1);
+#endif  /* _WIN32 */
+  if (n != 1)
+  {
+    DDS_WARNING("ddsrt_event_queue: read failed on trigger pipe\n");
     return DDS_RETCODE_ERROR;
-#endif /* _WIN32 */
-
+  }
 #endif /* !LWIP_SOCKET */
+
   return DDS_RETCODE_OK;
 }
 
-dds_return_t ddsrt_event_queue_add(ddsrt_event_queue_t* queue, ddsrt_event_t* evt) {
-  assert(queue);
-  assert(evt);
-
+int ddsrt_event_queue_add(ddsrt_event_queue_t* queue, ddsrt_event_t* evt)
+{
   ddsrt_mutex_lock(&queue->lock);
-  if (queue->nevents == queue->cevents) {
+  for (size_t i = 0; i < queue->nevents; i++)
+  {
+    if (queue->events[i] == evt)
+    {
+      ddsrt_mutex_unlock(&queue->lock);
+      return 0;
+    }
+  }
+
+  if (queue->nevents == queue->cevents)
+  {
     queue->cevents += EVENTS_CONTAINER_DELTA;
     ddsrt_realloc(queue->events, sizeof(ddsrt_event_t*) * queue->cevents);
   }
   
   queue->events[queue->nevents++] = evt;
   ddsrt_mutex_unlock(&queue->lock);
-  
-  return DDS_RETCODE_OK;
+  return 1;
 }
 
-dds_return_t ddsrt_event_queue_remove(ddsrt_event_queue_t* queue, ddsrt_event_t* evt) {
+void ddsrt_event_queue_trim(ddsrt_event_queue_t* queue, size_t entries)
+{
   ddsrt_mutex_lock(&queue->lock);
-  for (size_t i = 0; i < queue->nevents; i++) {
-    if (queue->events[i] == evt) {
+  if (entries < queue->nevents)
+    queue->nevents = entries;
+  ddsrt_mutex_unlock(&queue->lock);
+}
+
+dds_return_t ddsrt_event_queue_remove(ddsrt_event_queue_t* queue, ddsrt_event_t* evt)
+{
+  dds_return_t ret = DDS_RETCODE_ALREADY_DELETED;
+  ddsrt_mutex_lock(&queue->lock);
+  for (size_t i = 0; i < queue->nevents; i++)
+  {
+    if (queue->events[i] == evt)
+    {
       memmove(queue->events + i, queue->events + i + 1, (queue->nevents - i - 1)*sizeof(ddsrt_event_t*));
```

Comment:
Wouldn't it be possible to move the last event into the slot occupied by the removed one (like in the `kqueue` variant)?

Reply:
I was a little more careful in this implementation to conserve the ordering of the events.
I will look into whether that is still necessary.
If the ordering is no longer of importance (if the trim function is moved to a filter type), then yes, we can replace the wholesale copy of large parts of data with just switching the entry to be dropped with one "beyond" the array.